corgi/tasks/managed_services.py

```diff
@@ -0,0 +1,83 @@
+from django.utils import timezone
+
+from config.celery import app
+from corgi.collectors.app_interface import AppInterface
+from corgi.collectors.syft import Syft
+from corgi.core.models import Component, ComponentNode, ProductStream, SoftwareBuild
+from corgi.tasks.common import RETRY_KWARGS, RETRYABLE_ERRORS
+from corgi.tasks.sca import save_component
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def refresh_service_manifests():
+    services = ProductStream.objects.filter(meta_attr__managed_service_components__isnull=False)
+    service_metadata = AppInterface.fetch_service_metadata(services)
+
+    for service, components in service_metadata.items():
+        manifest_service(service.pk, components)
+        manifest_service.delay(service.pk, components)
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+@transaction.atomic
+def manifest_service(product_stream_id: str, service_components: list):
+    service = ProductStream.objects.get(pk=product_stream_id)
+
+    touched_build_ids = []
+    now = timezone.now()
+    for service_component in service_components:
+        analyzed_components = []
+        quay_repo = service_component.get("quay_repo_name")
+        if quay_repo:
+            quay_repo_full = f"quay.io/{quay_repo}"
+            component_data, scan_source = Syft.scan_repo_image(target_image=quay_repo_full)
+            for scanned_component in component_data:
+                scanned_component["meta"][
+                    "quay_image"
+                ] = f"{quay_repo_full}@{scan_source['target']['imageID']}"
+            analyzed_components.extend(component_data)
+
+        git_repo = service_component.get("git_repo_url")
+        if git_repo:
+            component_data, source_ref = Syft.scan_git_repo(target_url=git_repo)
+            for scanned_component in component_data:
+                scanned_component["meta"]["git_repo"] = f"{git_repo}@{source_ref}"
+            analyzed_components.extend(component_data)
+
+        if not analyzed_components:
+            continue
+
+        faux_version = now.strftime("%Y%m%d%H%M%S")
+        build = SoftwareBuild.objects.create(
+            name=service_component["name"],
+            build_type=SoftwareBuild.Type.APP_INTERFACE,
+            build_id=faux_version,
+            completion_time=now,
+            meta_attr={"service": service.name},
+        )
+        touched_build_ids.append(build.pk)
+
+        root_component, _ = Component.objects.update_or_create(
+            type=Component.Type.CONTAINER_IMAGE if quay_repo else Component.Type.GENERIC,
+            name=service_component["name"],
+            release="",
+            arch="",
+            defaults={
+                "version": faux_version,
+                "namespace": Component.Namespace.REDHAT,
+                "software_build": build,
+            },
+        )
+        root_component.productstreams.add(service)
```

Comment:
@jasinner @juspence I need some help here. What's the magic way to add a ProductStream to a component and then have it magically update the entire product taxonomy with the rest of the product-related models (Product, ProductVersion) based off of the Stream? I tried calling save_product_taxonomy on the Component but that complained it must be done on the build, but that doesn't seem to populate any data for all of the component of that build?